core/src/main/java/tech/pegasys/web3signer/core/service/http/handlers/keymanager/imports/ImportKeystoresHandler.java

```diff
@@ -141,35 +137,63 @@ public void handle(final RoutingContext context) {
             new ByteArrayInputStream(
                 parsedBody.getSlashingProtection().getBytes(StandardCharsets.UTF_8));
         slashingProtection.get().importDataWithFilter(slashingProtectionData, nonLoadedPubkeys);
-      } catch (Exception e) {
+      } catch (final Exception e) {
         context.fail(BAD_REQUEST, e);
         return;
       }
     }
 
-    final List<ImportKeystoreResult> results = new ArrayList<>();
+    final List<ImportKeystoreData> duplicateData = new ArrayList<>();
+    final List<ImportKeystoreData> toBeImportedData = new ArrayList<>();
     for (int i = 0; i < parsedBody.getKeystores().size(); i++) {
-      final String pubkey = pubkeysToImport.get(i);
-      try {
-        final String jsonKeystoreData = parsedBody.getKeystores().get(i);
-        final String password = parsedBody.getPasswords().get(i);
-
-        if (existingPubkeys.contains(pubkey)) {
-          // keystore already loaded
-          results.add(new ImportKeystoreResult(ImportKeystoreStatus.DUPLICATE, null));
-        } else {
-          validatorManager.addValidator(Bytes.fromHexString(pubkey), jsonKeystoreData, password);
-          results.add(new ImportKeystoreResult(ImportKeystoreStatus.IMPORTED, null));
-        }
-      } catch (final Exception e) {
-        // cleanup the current key being processed and continue
-        removeSignersAndCleanupImportedKeystoreFiles(List.of(pubkey));
-        results.add(
-            new ImportKeystoreResult(
-                ImportKeystoreStatus.ERROR, "Error importing keystore: " + e.getMessage()));
+      final String jsonKeystoreData = parsedBody.getKeystores().get(i);
+      final String password = parsedBody.getPasswords().get(i);
+      final String pubkey = getNormalizedPubKey(jsonKeystoreData);
+      if (existingPubkeys.contains(pubkey)) {
+        // keystore already loaded
+        duplicateData.add(
+            new ImportKeystoreData(
+                i, pubkey, jsonKeystoreData, password, new ImportKeystoreResult(DUPLICATE)));
+      } else {
+        toBeImportedData.add(
+            new ImportKeystoreData(
+                i, pubkey, jsonKeystoreData, password, new ImportKeystoreResult(IMPORTED)));
       }
     }
 
+    // process TO_BE_IMPORTED in parallel and add to validator manager. Change status to ERROR if
+    // any error happens.
+    final List<ImportKeystoreData> importedData =
```

Comment:
It's a shame we need to iterate over the list multiple times. You could probably do a zip using mapToObj and an IntStream to just over the stream once.

i.e. something like
```
IntStream.range(0, parsedBody.getKeystores().size())
            .mapToObj(i -> { create import keystore result; })
```